news/2022/09/21/compile_time.md

```diff
@@ -0,0 +1,1557 @@
+    du[3] = u[1]*u[2] - (8/3)*u[3]
+end
+u0 = [1.0;0.0;0.0]; tspan = (0.0,100.0)
+prob = ODEProblem(lorenz,u0,tspan)
+solve(prob,Rodas5())
+```
+
+In this case, the user provides us a model in the form of a function `lorenz`. This function is a mutating
+function, and from the ODE definition we have that `u0 isa Vector{Float64}`, `eltype(tspan) isa Float64`,
+and there are no parameters (and thus `typeof(p) isa SciMLBase.NullParameters`). From these facts we
+know that internal to the ODE solver when automatic differentiation is not being used, the type of `u`
+and the type of `du` match the `typeof(u0)` (we can also deduce the types required for automatic 
+differentiation, but that's a longer story which I will leave for the appendix). Therefore it is at the
+point of the `ODEProblem` construction that we have all of the information to do the function wrapping.
+
+We can thus have the `ODEProblem` call itself specialize on the input function `lorenz`, but immediately
+compute the wrapper as follows:
+
+```julia
+# Since the function mutates, make a wrapper that always throws away
+# the return and gives nothing, just in case the user accidentally forgets!
+struct Void{F}
+  f::F
+end
+
+function (f::Void{F})(x) where F
+  f.f(x)
+  nothing
+end
+
+FunctionWrapper{Nothing,Tuple{Vector{Float64},Vector{Float64},SciMLBase.NullParameters,Float64}}(Void(lorenz))
+```
+
+Thus the very first method of the call stack will have to recompile for every new ODE, but that's a trivial
+<100 microseconds call short call. All of the real functionality is then behind the next call, like:
+
+```julia
+function ODEProblem(f,u0,tspan,p;kwargs...)
+  _ODEProblem(wrapfunction(f),u0,tspan,p;kwargs...)
+end
+```
+
+so then all of the "real work" is precompiled. If this is done, then if `f` changes the `typeof(prob)`
+stays constant, and thus `solve` can be fully precompiled. In SciML we called this 
+[`SciMLBase.FunctionWrapperSpecialize`](https://scimlbase.sciml.ai/dev/interfaces/Problems/#Specialization-Levels).
+However, the difficulty with this form is that we have to commit to the `FunctionWrapper` very early.
+If `ODEProblem` is then attempted to be solved with some new solver that uses some new automatic
+differentiation technique, it may break when it sees the `FunctionWrapper`, so you may need to manually
+unwrap (`SciMLBase.unwrapped_f(prob.f)`) and it becomes a nightmare to maintain. Thus the real question
+is, how much specialization do you really need to avoid?
+
+Since the vast majority (`>99%`) of the compile time lives in the `solve(prob,Rodas5())` call, avoiding
+respecializing the rest of the `ODEProblem` call was simply over-engineering. Thus we found that a similar
+strategy could hold in the `solve` call itself. This looks like:
+
+```julia
+function solve(prob,alg)
+  # Pseudocode
+  if alg is okay with having the function wrapped
+    _prob = wrapped_f_prob(prob)
+  else
+    _prob = prob
+  end
+
+  __solve(prob,alg)

```

Comment:
Should probably be `__solve(_prob,alg)`?

Reply:
yes thanks